// sveld/lib/writer/writer-markdown.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = writeMarkdown;
const node_path_1 = require("node:path");
const WriterMarkdown_1 = __importDefault(require("./WriterMarkdown"));
const writer_ts_definitions_core_1 = require("./writer-ts-definitions-core");
const PROP_TABLE_HEADER = "| Prop name | Required | Kind | Reactive | Type | Default value | Description |\n| :- | :- | :- | :- |\n";
const SLOT_TABLE_HEADER = "| Slot name | Default | Props | Fallback |\n| :- | :- | :- | :- |\n";
const EVENT_TABLE_HEADER = "| Event name | Type | Detail | Description |\n| :- | :- | :- | :- |\n";
const MD_TYPE_UNDEFINED = "--";
const PIPE_REGEX = /\|/g;
const LT_REGEX = /</g;
const GT_REGEX = />/g;
const BACKTICK_REGEX = /`/g;
const NEWLINE_REGEX = /\n/g;
function formatPropType(type) {
    if (type === undefined)
        return MD_TYPE_UNDEFINED;
    return `<code>${type.replace(PIPE_REGEX, "&#124;")}</code>`;
}
function escapeHtml(text) {
    return text.replace(LT_REGEX, "&lt;").replace(GT_REGEX, "&gt;");
}
function formatPropValue(value) {
    if (value === undefined)
        return MD_TYPE_UNDEFINED;
    return `<code>${value.replace(BACKTICK_REGEX, "\\`").replace(PIPE_REGEX, "&#124;")}</code>`;
}
function formatPropDescription(description) {
    if (description === undefined || description.trim().length === 0)
        return MD_TYPE_UNDEFINED;
    return escapeHtml(description).replace(NEWLINE_REGEX, "<br />");
}
function formatSlotProps(props) {
    if (props === undefined || props === "{}")
        return MD_TYPE_UNDEFINED;
    return formatPropType((0, writer_ts_definitions_core_1.formatTsProps)(props).replace(NEWLINE_REGEX, " "));
}
function formatSlotFallback(fallback) {
    if (fallback === undefined)
        return MD_TYPE_UNDEFINED;
    return formatPropType(escapeHtml(fallback).replace(NEWLINE_REGEX, "<br />"));
}
function formatEventDetail(detail) {
    if (detail === undefined)
        return MD_TYPE_UNDEFINED;
    return formatPropType(detail.replace(NEWLINE_REGEX, " "));
}
async function writeMarkdown(components, options) {
    const write = options?.write !== false;
    const document = new WriterMarkdown_1.default({
        onAppend: (type, document) => {
            options.onAppend?.call(null, type, document, components);
        },
    });
    document.append("h1", "Component Index");
    document.append("h2", "Components").tableOfContents();
    document.append("divider");
    const keys = Array.from(components.keys()).sort();
    for (const key of keys) {
        const component = components.get(key);
        if (!component)
            continue;
        document.append("h2", `\`${component.moduleName}\``);
        if (component.typedefs.length > 0) {
            document.append("h3", "Types").append("raw", `\`\`\`ts\n${(0, writer_ts_definitions_core_1.getTypeDefs)({
                typedefs: component.typedefs,
            })}\n\`\`\`\n\n`);
        }
        document.append("h3", "Props");
        if (component.props.length > 0) {
            document.append("raw", PROP_TABLE_HEADER);
            const sortedProps = [...component.props].sort((a) => {
                if (a.reactive)
                    return -1;
                if (a.constant)
                    return 1;
                return 0;
            });
            for (const prop of sortedProps) {
                document.append("raw", `| ${prop.name} | ${prop.isRequired ? "Yes" : "No"} | ${`<code>${prop.kind}</code>`} | ${prop.reactive ? "Yes" : "No"} | ${formatPropType(prop.type)} | ${formatPropValue(prop.value)} | ${formatPropDescription(prop.description)} |\n`);
            }
        }
        else {
            document.append("p", "None.");
        }
        document.append("h3", "Slots");
        if (component.slots.length > 0) {
            document.append("raw", SLOT_TABLE_HEADER);
            for (const slot of component.slots) {
                document.append("raw", `| ${slot.default ? MD_TYPE_UNDEFINED : slot.name} | ${slot.default ? "Yes" : "No"} | ${formatSlotProps(slot.slot_props)} | ${formatSlotFallback(slot.fallback)} |\n`);
            }
        }
        else {
            document.append("p", "None.");
        }
        document.append("h3", "Events");
        if (component.events.length > 0) {
            document.append("raw", EVENT_TABLE_HEADER);
            for (const event of component.events) {
                document.append("raw", `| ${event.name} | ${event.type} | ${event.type === "dispatched" ? formatEventDetail(event.detail) : MD_TYPE_UNDEFINED} | ${formatPropDescription(event.description)} |\n`);
            }
        }
        else {
            document.append("p", "None.");
        }
    }
    if (write) {
        const outFile = (0, node_path_1.join)(process.cwd(), options.outFile);
        await document.write(outFile, document.end());
        console.log(`created "${options.outFile}".`);
    }
    return document.end();
}
