// sveld/lib/writer/writer-json.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = writeJson;
const node_path_1 = __importDefault(require("node:path"));
const path_1 = require("../path");
const Writer_1 = __importDefault(require("./Writer"));
async function writeJsonComponents(components, options) {
    const output = Array.from(components, ([_moduleName, component]) => ({
        ...component,
        filePath: (0, path_1.normalizeSeparators)(node_path_1.default.join(options.inputDir, node_path_1.default.normalize(component.filePath))),
    })).sort((a, b) => {
        if (a.moduleName < b.moduleName)
            return -1;
        if (a.moduleName > b.moduleName)
            return 1;
        return 0;
    });
    output.map((c) => {
        const outFile = node_path_1.default.resolve(node_path_1.default.join(options.outDir || "", `${c.moduleName}.api.json`));
        const writer = new Writer_1.default({ parser: "json", printWidth: 80 });
        console.log(`created ${outFile}"\n`);
        return writer.write(outFile, JSON.stringify(c));
    });
}
async function writeJsonLocal(components, options) {
    const output = {
        total: components.size,
        components: Array.from(components, ([_moduleName, component]) => ({
            ...component,
            filePath: (0, path_1.normalizeSeparators)(node_path_1.default.join(options.inputDir, node_path_1.default.normalize(component.filePath))),
        })).sort((a, b) => {
            if (a.moduleName < b.moduleName)
                return -1;
            if (a.moduleName > b.moduleName)
                return 1;
            return 0;
        }),
    };
    const output_path = node_path_1.default.join(process.cwd(), options.outFile);
    const writer = new Writer_1.default({ parser: "json", printWidth: 80 });
    await writer.write(output_path, JSON.stringify(output));
    console.log(`created "${options.outFile}".\n`);
}
async function writeJson(components, options) {
    if (options.outDir) {
        await writeJsonComponents(components, options);
    }
    else {
        await writeJsonLocal(components, options);
    }
}
