// sveld/lib/writer/WriterMarkdown.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Writer_1 = __importDefault(require("./Writer"));
const BACKTICK_REGEX = /`/g;
const WHITESPACE_REGEX = /\s+/g;
class WriterMarkdown extends Writer_1.default {
    onAppend;
    source = "";
    hasToC = false;
    toc = [];
    constructor(options) {
        super({ parser: "markdown", printWidth: 80 });
        this.onAppend = options.onAppend;
    }
    appendLineBreaks() {
        this.source += "\n\n";
        return this;
    }
    append(type, raw) {
        switch (type) {
            case "h1":
            case "h2":
            case "h3":
            case "h4":
            case "h5":
            case "h6": {
                const length = Number(type.slice(-1));
                this.source += `${Array.from({ length })
                    .map((_) => "#")
                    .join("")} ${raw}`;
                if (this.hasToC && type === "h2") {
                    this.toc.push({
                        array: Array.from({ length: (length - 1) * 2 }),
                        raw: raw ?? "",
                    });
                }
                break;
            }
            case "quote":
                this.source += `> ${raw}`;
                break;
            case "p":
                this.source += raw;
                break;
            case "divider":
                this.source += "---";
                break;
            case "raw":
                this.source += raw;
                break;
        }
        if (type !== "raw")
            this.appendLineBreaks();
        this.onAppend?.call(this, type, this);
        return this;
    }
    tableOfContents() {
        this.source += "<!-- __TOC__ -->";
        this.hasToC = true;
        this.appendLineBreaks();
        return this;
    }
    end() {
        this.source = this.source.replace("<!-- __TOC__ -->", this.toc
            .map(({ array, raw }) => {
            return `${array.join(" ")} - [${raw}](#${raw.toLowerCase().replace(BACKTICK_REGEX, "").replace(WHITESPACE_REGEX, "-")})`;
        })
            .join("\n"));
        return this.source;
    }
}
exports.default = WriterMarkdown;
