// sveld/lib/ComponentParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// TODO: upgrading to Svelte 4 shows a lot of TS errors. Ignore for now but resolve.
// @ts-nocheck
const comment_parser_1 = require("comment-parser");
const compiler_1 = require("svelte/compiler");
const element_tag_map_1 = require("./element-tag-map");
const VAR_DECLARATION_REGEX = /(?:const|let|function)\s+(\w+)\s*[=(]/;
const DEFAULT_SLOT_NAME = null;
const TYPEDEF_END_REGEX = /(\}|\};)$/;
const PROPERTY_DESCRIPTION_REGEX = /^-\s*/;
const CONTEXT_KEY_SPLIT_REGEX = /[-_\s]+/;
const COMPONENT_COMMENT_REGEX = /^@component/;
const CARRIAGE_RETURN_REGEX = /\r/g;
const NEWLINE_CR_REGEX = /[\r\n]+/g;
class ComponentParser {
    options;
    source;
    compiled;
    parsed;
    rest_props;
    extends;
    componentComment;
    reactive_vars = new Set();
    vars = new Set();
    props = new Map();
    moduleExports = new Map();
    slots = new Map();
    events = new Map();
    eventDescriptions = new Map();
    forwardedEvents = new Map();
    typedefs = new Map();
    generics = null;
    bindings = new Map();
    contexts = new Map();
    variableInfoCache = new Map();
    constructor(options) {
        this.options = options;
    }
    static mapToArray(map) {
        return Array.from(map, ([_key, value]) => value);
    }
    static assignValue(value) {
        return value === undefined || value === "" ? undefined : value;
    }
    static formatComment(comment) {
        let formatted_comment = comment;
        if (!formatted_comment.startsWith("/*")) {
            formatted_comment = `/*${formatted_comment}`;
        }
        if (!formatted_comment.endsWith("*/")) {
            formatted_comment += "*/";
        }
        return formatted_comment;
    }
    /**
     * Finds the last comment from an array of leading comments.
     * TypeScript directives are stripped before parsing, so we can safely take the last comment.
     */
    static findJSDocComment(leadingComments) {
        if (!leadingComments || leadingComments.length === 0)
            return undefined;
        return leadingComments[leadingComments.length - 1];
    }
    /**
     * Check if a MemberExpression represents a well-known numeric constant.
     * (e.g., Number.POSITIVE_INFINITY, Math.PI, etc.)
     */
    isNumericConstant(memberExpr) {
        if (memberExpr.type !== "MemberExpression")
            return false;
        const objectName = memberExpr.object?.name;
        const propertyName = memberExpr.property?.name;
        if (!objectName || !propertyName)
            return false;
        if (objectName === "Number") {
            return [
                "POSITIVE_INFINITY",
                "NEGATIVE_INFINITY",
                "MAX_VALUE",
                "MIN_VALUE",
                "MAX_SAFE_INTEGER",
                "MIN_SAFE_INTEGER",
                "EPSILON",
                "NaN",
            ].includes(propertyName);
        }
        if (objectName === "Math") {
            return ["PI", "E", "LN2", "LN10", "LOG2E", "LOG10E", "SQRT2", "SQRT1_2"].includes(propertyName);
        }
        return false;
    }
    sourceAtPos(start, end) {
        return this.source?.slice(start, end);
    }
    collectReactiveVars() {
        const reactiveVars = this.compiled?.vars.filter(({ reassigned, writable }) => reassigned && writable) ?? [];
        for (const { name } of reactiveVars) {
            this.reactive_vars.add(name);
        }
    }
    addProp(prop_name, data) {
        if (ComponentParser.assignValue(prop_name) === undefined)
            return;
        if (this.props.has(prop_name)) {
            const existing_slot = this.props.get(prop_name);
            this.props.set(prop_name, {
                ...existing_slot,
                ...data,
            });
        }
        else {
            this.props.set(prop_name, data);
        }
    }
    addModuleExport(prop_name, data) {
        if (ComponentParser.assignValue(prop_name) === undefined)
            return;
        if (this.moduleExports.has(prop_name)) {
            const existing_slot = this.moduleExports.get(prop_name);
            this.moduleExports.set(prop_name, {
                ...existing_slot,
                ...data,
            });
        }
        else {
            this.moduleExports.set(prop_name, data);
        }
    }
    aliasType(type) {
        if (type === "*")
            return "any";
        return type;
    }
    addSlot({ slot_name, slot_props, slot_fallback, slot_description, }) {
        const default_slot = slot_name === undefined || slot_name === "";
        const name = default_slot ? DEFAULT_SLOT_NAME : (slot_name ?? "");
        const fallback = ComponentParser.assignValue(slot_fallback);
        const props = ComponentParser.assignValue(slot_props);
        const description = slot_description?.split("-").pop()?.trim();
        if (this.slots.has(name)) {
            const existing_slot = this.slots.get(name);
            this.slots.set(name, {
                ...existing_slot,
                fallback,
                slot_props: existing_slot.slot_props === undefined ? props : existing_slot.slot_props,
                description: existing_slot.description || description,
            });
        }
        else {
            this.slots.set(name, {
                name,
                default: default_slot,
                fallback,
                slot_props,
                description,
            });
        }
    }
    addDispatchedEvent({ name, detail, has_argument, description, }) {
        if (name === undefined)
            return;
        /**
         * `e.detail` should be `null` if the dispatcher
         * is not provided a second argument and if
         * `@event` is not specified.
         */
        const default_detail = !has_argument && !detail ? "null" : ComponentParser.assignValue(detail);
        const event_description = description?.split("-").pop()?.trim();
        if (this.events.has(name)) {
            const existing_event = this.events.get(name);
            this.events.set(name, {
                ...existing_event,
                detail: existing_event.detail === undefined ? default_detail : existing_event.detail,
                description: existing_event.description || event_description,
            });
        }
        else {
            this.events.set(name, {
                type: "dispatched",
                name,
                detail: default_detail,
                description: event_description,
            });
        }
    }
    parseCustomTypes() {
        for (const { tags, description: commentDescription } of (0, comment_parser_1.parse)(this.source, { spacing: "preserve" })) {
            let currentEventName;
            let currentEventType;
            let currentEventDescription;
            const eventProperties = [];
            let currentTypedefName;
            let currentTypedefType;
            let currentTypedefDescription;
            const typedefProperties = [];
            const finalizeEvent = () => {
                if (currentEventName !== undefined) {
                    let detailType;
                    if (eventProperties.length > 0) {
                        detailType = this.buildEventDetailFromProperties(eventProperties, currentEventName);
                    }
                    else {
                        detailType = currentEventType || "";
                    }
                    this.addDispatchedEvent({
                        name: currentEventName,
                        detail: detailType,
                        has_argument: false,
                        description: currentEventDescription,
                    });
                    this.eventDescriptions.set(currentEventName, currentEventDescription);
                    eventProperties.length = 0;
                    currentEventName = undefined;
                    currentEventType = undefined;
                    currentEventDescription = undefined;
                }
            };
            const finalizeTypedef = () => {
                if (currentTypedefName !== undefined) {
                    let typedefType;
                    let typedefTs;
                    if (typedefProperties.length > 0) {
                        // Build type alias with property descriptions
                        typedefType = this.buildEventDetailFromProperties(typedefProperties);
                        typedefTs = `type ${currentTypedefName} = ${typedefType}`;
                    }
                    else if (currentTypedefType) {
                        // Use inline type definition (existing behavior)
                        typedefType = currentTypedefType;
                        typedefTs = TYPEDEF_END_REGEX.test(typedefType)
                            ? `interface ${currentTypedefName} ${typedefType}`
                            : `type ${currentTypedefName} = ${typedefType}`;
                    }
                    else {
                        // No type or properties specified, default to empty object
                        typedefType = "{}";
                        typedefTs = `type ${currentTypedefName} = ${typedefType}`;
                    }
                    this.typedefs.set(currentTypedefName, {
                        type: typedefType,
                        name: currentTypedefName,
                        description: ComponentParser.assignValue(currentTypedefDescription),
                        ts: typedefTs,
                    });
                    typedefProperties.length = 0;
                    currentTypedefName = undefined;
                    currentTypedefType = undefined;
                    currentTypedefDescription = undefined;
                }
            };
            for (const { tag, type: tagType, name, description, optional, default: defaultValue } of tags) {
                const type = this.aliasType(tagType);
                switch (tag) {
                    case "extends":
                        this.extends = {
                            interface: name,
                            import: type,
                        };
                        break;
                    case "restProps":
                        this.rest_props = {
                            type: "Element",
                            name: type,
                        };
                        break;
                    case "slot":
                        this.addSlot({
                            slot_name: name,
                            slot_props: type,
                            slot_description: description ? description : undefined,
                        });
                        break;
                    case "event":
                        // Finalize any previous event being built
                        finalizeEvent();
                        // Start tracking new event
                        currentEventName = name;
                        currentEventType = type;
                        // Use the main comment description if available, otherwise use inline description
                        currentEventDescription = commentDescription?.trim() || description || undefined;
                        break;
                    case "type":
                        // Track the @type tag for the current event
                        if (currentEventName !== undefined) {
                            currentEventType = type;
                        }
                        break;
                    case "property": {
                        // Collect properties for the current event or typedef
                        const propertyData = {
                            name,
                            type,
                            description: description?.replace(PROPERTY_DESCRIPTION_REGEX, "").trim(),
                            optional: optional || false,
                            default: defaultValue,
                        };
                        if (currentEventName !== undefined) {
                            eventProperties.push(propertyData);
                        }
                        else if (currentTypedefName !== undefined) {
                            typedefProperties.push(propertyData);
                        }
                        break;
                    }
                    case "typedef": {
                        // Finalize any previous typedef being built
                        finalizeTypedef();
                        // Start tracking new typedef
                        currentTypedefName = name;
                        currentTypedefType = type;
                        // Use inline description if present, otherwise use comment description
                        const trimmedCommentDesc = commentDescription?.trim();
                        currentTypedefDescription =
                            description || (trimmedCommentDesc && trimmedCommentDesc !== "}" ? trimmedCommentDesc : undefined);
                        break;
                    }
                    case "generics":
                        this.generics = [name, type];
                        break;
                }
            }
            // Finalize any remaining event or typedef
            finalizeEvent();
            finalizeTypedef();
        }
    }
    buildEventDetailFromProperties(properties) {
        if (properties.length === 0)
            return "null";
        // Build inline object type with property descriptions as JSDoc comments
        const props = properties
            .map(({ name, type, description, optional, default: defaultValue }) => {
            const optionalMarker = optional ? "?" : "";
            let comment = description || "";
            // Add default value to description if present
            if (defaultValue && comment) {
                comment = `${comment} @default ${defaultValue}`;
            }
            else if (defaultValue) {
                comment = `@default ${defaultValue}`;
            }
            if (comment) {
                return `/** ${comment} */ ${name}${optionalMarker}: ${type};`;
            }
            return `${name}${optionalMarker}: ${type};`;
        })
            .join(" ");
        return `{ ${props} }`;
    }
    generateContextTypeName(key) {
        // Convert "simple-modal" -> "SimpleModalContext"
        // Convert "Tabs" -> "TabsContext"
        const parts = key.split(CONTEXT_KEY_SPLIT_REGEX);
        const capitalized = parts.map((p) => p.charAt(0).toUpperCase() + p.slice(1)).join("");
        return `${capitalized}Context`;
    }
    buildVariableInfoCache() {
        if (!this.source)
            return;
        const lines = this.source.split("\n");
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i].trim();
            // Match variable declarations
            const varMatch = line.match(VAR_DECLARATION_REGEX);
            if (varMatch) {
                const varName = varMatch[1];
                // Look backwards for JSDoc comment
                for (let j = i - 1; j >= 0; j--) {
                    const prevLine = lines[j].trim();
                    // Stop if we hit a non-comment, non-empty line
                    if (prevLine && !prevLine.startsWith("*") && !prevLine.startsWith("/*") && !prevLine.startsWith("//")) {
                        break;
                    }
                    // Found start of JSDoc comment
                    if (prevLine.startsWith("/**")) {
                        // Extract the JSDoc comment block
                        const commentLines = [];
                        for (let k = j; k < i; k++) {
                            commentLines.push(lines[k]);
                        }
                        const commentBlock = commentLines.join("\n");
                        // Parse the JSDoc
                        const parsed = (0, comment_parser_1.parse)(commentBlock, { spacing: "preserve" });
                        if (parsed[0]?.tags) {
                            const typeTag = parsed[0].tags.find((t) => t.tag === "type");
                            if (typeTag) {
                                this.variableInfoCache.set(varName, {
                                    type: this.aliasType(typeTag.type),
                                    description: parsed[0].description || typeTag.description,
                                });
                            }
                        }
                        break;
                    }
                }
            }
        }
    }
    findVariableTypeAndDescription(varName) {
        const cached = this.variableInfoCache.get(varName);
        if (cached) {
            return cached;
        }
        // Search through the source code directly for JSDoc comments
        if (!this.source)
            return null;
        // Build a map of variable names to their types by looking at the source
        const lines = this.source.split("\n");
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i].trim();
            // Check if this line declares our variable
            // Match patterns like: const varName = ..., let varName = ..., function varName
            const constMatch = line.match(new RegExp(`const\\s+${varName}\\s*=`));
            const letMatch = line.match(new RegExp(`let\\s+${varName}\\s*=`));
            const funcMatch = line.match(new RegExp(`function\\s+${varName}\\s*\\(`));
            if (constMatch || letMatch || funcMatch) {
                // Look backwards for JSDoc comment
                for (let j = i - 1; j >= 0; j--) {
                    const prevLine = lines[j].trim();
                    // Stop if we hit a non-comment, non-empty line
                    if (prevLine && !prevLine.startsWith("*") && !prevLine.startsWith("/*") && !prevLine.startsWith("//")) {
                        break;
                    }
                    // Found start of JSDoc comment
                    if (prevLine.startsWith("/**")) {
                        // Extract the JSDoc comment block
                        const commentLines = [];
                        for (let k = j; k < i; k++) {
                            commentLines.push(lines[k]);
                        }
                        const commentBlock = commentLines.join("\n");
                        // Parse the JSDoc
                        const parsed = (0, comment_parser_1.parse)(commentBlock, { spacing: "preserve" });
                        if (parsed[0]?.tags) {
                            const typeTag = parsed[0].tags.find((t) => t.tag === "type");
                            if (typeTag) {
                                return {
                                    type: this.aliasType(typeTag.type),
                                    description: parsed[0].description || typeTag.description,
                                };
                            }
                        }
                        break;
                    }
                }
            }
        }
        return null;
    }
    parseContextValue(node, key) {
        if (node.type === "ObjectExpression") {
            // Parse object literal: { open, close }
            const properties = [];
            for (const prop of node.properties) {
                if (prop.type === "Property" || prop.type === "ObjectProperty") {
                    const propName = prop.key.name || prop.key.value;
                    // Try to find the variable definition to get its JSDoc type
                    let propType = "any";
                    let propDescription;
                    if (prop.value.type === "Identifier") {
                        const varName = prop.value.name;
                        const varInfo = this.findVariableTypeAndDescription(varName);
                        if (varInfo) {
                            propType = varInfo.type;
                            propDescription = varInfo.description;
                        }
                        else if (this.options?.verbose) {
                            console.warn(`Warning: Context "${key}" property "${propName}" has no type annotation. Using "any".`);
                        }
                    }
                    else if (prop.value.type === "ArrowFunctionExpression" || prop.value.type === "FunctionExpression") {
                        // Inline function
                        const params = prop.value.params.map((p) => `${p.name || "arg"}: any`).join(", ");
                        propType = `(${params}) => any`;
                    }
                    else if (prop.value.type === "Literal") {
                        // Literal value
                        propType = typeof prop.value.value;
                    }
                    properties.push({
                        name: propName,
                        type: propType,
                        description: propDescription,
                        optional: false,
                    });
                }
            }
            return {
                key,
                typeName: this.generateContextTypeName(key),
                properties,
                description: undefined,
            };
        }
        else if (node.type === "Identifier") {
            // setContext('key', someVariable)
            const varName = node.name;
            const varInfo = this.findVariableTypeAndDescription(varName);
            if (varInfo) {
                return {
                    key,
                    typeName: this.generateContextTypeName(key),
                    properties: [
                        {
                            name: varName,
                            type: varInfo.type,
                            description: varInfo.description,
                            optional: false,
                        },
                    ],
                };
            }
            else if (this.options?.verbose) {
                console.warn(`Warning: Context "${key}" variable "${varName}" has no type annotation. Using "any".`);
            }
            // Still create context with 'any' type
            return {
                key,
                typeName: this.generateContextTypeName(key),
                properties: [
                    {
                        name: varName,
                        type: "any",
                        description: undefined,
                        optional: false,
                    },
                ],
            };
        }
        return null;
    }
    parseSetContextCall(node) {
        // Extract context key (first argument)
        const keyArg = node.arguments[0];
        if (!keyArg)
            return;
        let contextKey = null;
        if (keyArg.type === "Literal") {
            contextKey = keyArg.value;
        }
        else if (keyArg.type === "TemplateLiteral") {
            // Handle simple template literals
            if (keyArg.quasis?.length === 1) {
                contextKey = keyArg.quasis[0].value.cooked;
            }
            else if (this.options?.verbose) {
                console.warn("Warning: Skipping setContext with dynamic template literal key");
            }
        }
        else if (this.options?.verbose) {
            console.warn(`Warning: Skipping setContext with non-literal key (type: ${keyArg.type})`);
        }
        if (!contextKey)
            return;
        // Extract context value (second argument)
        const valueArg = node.arguments[1];
        if (!valueArg)
            return;
        // Parse the context object
        const contextInfo = this.parseContextValue(valueArg, contextKey);
        if (contextInfo) {
            // Check if context with same key already exists
            if (this.contexts.has(contextKey)) {
                if (this.options?.verbose) {
                    console.warn(`Warning: Multiple setContext calls with key "${contextKey}". Using first occurrence.`);
                }
            }
            else {
                this.contexts.set(contextKey, contextInfo);
            }
        }
    }
    cleanup() {
        this.source = undefined;
        this.compiled = undefined;
        this.parsed = undefined;
        this.rest_props = undefined;
        this.extends = undefined;
        this.componentComment = undefined;
        this.reactive_vars.clear();
        this.props.clear();
        this.moduleExports.clear();
        this.slots.clear();
        this.events.clear();
        this.eventDescriptions.clear();
        this.forwardedEvents.clear();
        this.typedefs.clear();
        this.generics = null;
        this.bindings.clear();
        this.contexts.clear();
        this.variableInfoCache.clear();
    }
    // Pre-compiled regexes for better performance
    static SCRIPT_BLOCK_REGEX = /(<script[^>]*>)([\s\S]*?)(<\/script>)/gi;
    static TS_DIRECTIVE_REGEX = /\/\/\s*@ts-[^\n\r]*/g;
    /**
     * Strips TypeScript directive comments from script blocks only.
     */
    static stripTypeScriptDirectivesFromScripts(source) {
        // Find all script blocks and strip directives only from within them
        // Note: Need to reset lastIndex for global regex
        ComponentParser.SCRIPT_BLOCK_REGEX.lastIndex = 0;
        return source.replace(ComponentParser.SCRIPT_BLOCK_REGEX, (_match, openTag, scriptContent, closeTag) => {
            // Remove TypeScript directives from script content only
            ComponentParser.TS_DIRECTIVE_REGEX.lastIndex = 0;
            const cleanedContent = scriptContent.replace(ComponentParser.TS_DIRECTIVE_REGEX, "");
            return openTag + cleanedContent + closeTag;
        });
    }
    parseSvelteComponent(source, diagnostics) {
        if (this.options?.verbose) {
            console.log(`[parsing] "${diagnostics.moduleName}" ${diagnostics.filePath}`);
        }
        this.cleanup();
        // Strip TypeScript directives from script blocks only to prevent interference with JSDoc
        const cleanedSource = ComponentParser.stripTypeScriptDirectivesFromScripts(source);
        this.source = cleanedSource;
        // Parse once - compile() internally calls parse(), so we can extract the AST from it
        const compiled = (0, compiler_1.compile)(cleanedSource);
        this.compiled = compiled;
        // Reuse the AST from compilation instead of parsing again
        // The compile result includes the parsed AST
        this.parsed = compiled.ast || (0, compiler_1.parse)(cleanedSource);
        this.collectReactiveVars();
        this.buildVariableInfoCache();
        this.parseCustomTypes();
        if (this.parsed?.module) {
            (0, compiler_1.walk)(this.parsed?.module, {
                enter: (node) => {
                    if (node.type === "ExportNamedDeclaration") {
                        // Skip re-exports (e.g., export { A, B } from 'library').
                        if (node.declaration == null) {
                            return;
                        }
                        const { type: declaration_type, id, init, } = node.declaration?.declarations ? node.declaration.declarations[0] : node.declaration;
                        const prop_name = id.name;
                        let value;
                        let type;
                        let kind = node.declaration.kind;
                        let description;
                        let isFunction = false;
                        let isFunctionDeclaration = false;
                        if (init != null) {
                            if (init.type === "ObjectExpression" ||
                                init.type === "BinaryExpression" ||
                                init.type === "ArrayExpression" ||
                                init.type === "ArrowFunctionExpression") {
                                value = this.sourceAtPos(init.start, init.end)?.replace(NEWLINE_CR_REGEX, " ");
                                type = value;
                                isFunction = init.type === "ArrowFunctionExpression";
                                if (init.type === "BinaryExpression") {
                                    if (init?.left.type === "Literal" && typeof init?.left.value === "string") {
                                        type = "string";
                                    }
                                }
                                // For arrow functions, use a generic function type instead of the implementation
                                // Don't store the implementation in value - it clutters documentation
                                if (init.type === "ArrowFunctionExpression") {
                                    type = "(...args: any[]) => any";
                                    value = undefined;
                                }
                            }
                            else {
                                if (init.type === "UnaryExpression") {
                                    value = this.sourceAtPos(init.start, init.end);
                                    type = typeof init.argument?.value;
                                }
                                else if (init.type === "Identifier") {
                                    // Handle non-literal defaults like variable references and global identifiers.
                                    // Don't infer type, just preserve existing type annotation.
                                    value = this.sourceAtPos(init.start, init.end);
                                }
                                else if (init.type === "MemberExpression") {
                                    // Handle member expressions like Number.POSITIVE_INFINITY, Math.PI, etc.
                                    value = this.sourceAtPos(init.start, init.end);
                                    // Infer type as "number" for well-known numeric constants
                                    if (this.isNumericConstant(init)) {
                                        type = "number";
                                    }
                                    // Otherwise, don't infer type, just preserve existing type annotation.
                                }
                                else {
                                    value = init.raw;
                                    type = init.value == null ? undefined : typeof init.value;
                                }
                            }
                        }
                        if (declaration_type === "FunctionDeclaration") {
                            // Don't store function body in value - it clutters documentation
                            // The type signature is what matters for API docs
                            value = undefined;
                            type = "() => any";
                            kind = "function";
                            isFunction = true;
                            isFunctionDeclaration = true;
                        }
                        let params;
                        let returnType;
                        if (node.leadingComments) {
                            const jsdoc_comment = ComponentParser.findJSDocComment(node.leadingComments);
                            if (jsdoc_comment) {
                                const comment = (0, comment_parser_1.parse)(ComponentParser.formatComment(jsdoc_comment.value), {
                                    spacing: "preserve",
                                });
                                // Extract @type tag
                                const typeTag = comment[0]?.tags.find((t) => t.tag === "type");
                                if (typeTag)
                                    type = this.aliasType(typeTag.type);
                                // Extract @param tags
                                const paramTags = comment[0]?.tags.filter((t) => t.tag === "param") ?? [];
                                if (paramTags.length > 0) {
                                    params = paramTags
                                        .filter((tag) => !tag.name.includes(".")) // Exclude nested params like "options.expand"
                                        .map((tag) => ({
                                        name: tag.name,
                                        type: this.aliasType(tag.type),
                                        description: tag.description?.replace(PROPERTY_DESCRIPTION_REGEX, "").trim(),
                                        optional: tag.optional || false,
                                    }));
                                }
                                // Extract @returns/@return tag
                                const returnsTag = comment[0]?.tags.find((t) => t.tag === "returns" || t.tag === "return");
                                if (returnsTag)
                                    returnType = this.aliasType(returnsTag.type);
                                // Build description from comment description and non-param/non-type tags
                                const commentDescription = ComponentParser.assignValue(comment[0]?.description?.trim());
                                const additionalTags = comment[0]?.tags.filter((tag) => ![
                                    "type",
                                    "param",
                                    "returns",
                                    "return",
                                    "extends",
                                    "restProps",
                                    "slot",
                                    "event",
                                    "typedef",
                                ].includes(tag.tag)) ?? [];
                                if (commentDescription || additionalTags.length > 0) {
                                    description = commentDescription || "";
                                    for (const tag of additionalTags) {
                                        description += `${description ? "\n" : ""}@${tag.tag}${tag.name ? ` ${tag.name}` : ""}${tag.description ? ` ${tag.description}` : ""}`;
                                    }
                                }
                            }
                        }
                        if (!description && this.typedefs.has(type)) {
                            description = this.typedefs.get(type)?.description;
                        }
                        this.addModuleExport(prop_name, {
                            name: prop_name,
                            kind,
                            description,
                            type,
                            value,
                            params,
                            returnType,
                            isFunction,
                            isFunctionDeclaration,
                            isRequired: false,
                            constant: kind === "const",
                            reactive: false,
                        });
                    }
                },
            });
        }
        let dispatcher_name;
        const callees = [];
        (0, compiler_1.walk)({ html: this.parsed.html, instance: this.parsed.instance }, {
            enter: (node, parent, _prop) => {
                if (node.type === "CallExpression") {
                    if (node.callee.name === "createEventDispatcher") {
                        dispatcher_name = parent?.id.name;
                    }
                    if (node.callee.name === "setContext") {
                        this.parseSetContextCall(node, parent);
                    }
                    callees.push({
                        name: node.callee.name,
                        arguments: node.arguments,
                    });
                }
                if (node.type === "Spread" && node?.expression.name === "$$restProps") {
                    if (this.rest_props === undefined && (parent?.type === "InlineComponent" || parent?.type === "Element")) {
                        const restProps = {
                            type: parent.type,
                            name: parent.name,
                        };
                        // Handle svelte:element - check if this attribute is hardcoded
                        if (parent.type === "Element" && parent.name === "svelte:element") {
                            // The 'this' value is stored in the 'tag' property of the Element node
                            // If tag is a string, it's hardcoded; if undefined/null, it's dynamic
                            if (typeof parent.tag === "string") {
                                restProps.thisValue = parent.tag;
                            }
                            // If tag is undefined or not a string, thisValue remains undefined (dynamic)
                        }
                        this.rest_props = restProps;
                    }
                }
                if (node.type === "VariableDeclaration") {
                    this.vars.add(node);
                }
                if (node.type === "ExportNamedDeclaration") {
                    // Handle export {}
                    if (node.declaration == null && node.specifiers.length === 0) {
                        return;
                    }
                    // Handle renamed exports
                    let prop_name;
                    if (node.declaration == null && node.specifiers[0]?.type === "ExportSpecifier") {
                        const specifier = node.specifiers[0];
                        const localName = specifier.local.name;
                        const exportedName = specifier.exported.name;
                        let declaration;
                        // Search through all variable declarations for this variable
                        //  Limitation: the variable must have been declared before the export
                        for (const varDecl of this.vars) {
                            if (varDecl.declarations.some((decl) => decl.id.type === "Identifier" && decl.id.name === localName)) {
                                declaration = varDecl;
                            }
                        }
                        node.declaration = declaration;
                        prop_name = exportedName;
                    }
                    // Skip re-exports (e.g., export { A, B } from 'library').
                    if (node.declaration == null) {
                        return;
                    }
                    const { type: declaration_type, id, init, } = node.declaration.declarations ? node.declaration.declarations[0] : node.declaration;
                    prop_name ??= id.name;
                    let value;
                    let type;
                    let kind = node.declaration.kind;
                    let description;
                    let isFunction = false;
                    let isFunctionDeclaration = false;
                    const isRequired = kind === "let" && init == null;
                    if (init != null) {
                        if (init.type === "ObjectExpression" ||
                            init.type === "BinaryExpression" ||
                            init.type === "ArrayExpression" ||
                            init.type === "ArrowFunctionExpression") {
                            value = this.sourceAtPos(init.start, init.end)?.replace(NEWLINE_CR_REGEX, " ");
                            type = value;
                            isFunction = init.type === "ArrowFunctionExpression";
                            if (init.type === "BinaryExpression") {
                                if (init?.left.type === "Literal" && typeof init?.left.value === "string") {
                                    type = "string";
                                }
                            }
                            // For arrow functions, use a generic function type instead of the implementation
                            // Don't store the implementation in value - it clutters documentation
                            if (init.type === "ArrowFunctionExpression") {
                                type = "(...args: any[]) => any";
                                value = undefined;
                            }
                        }
                        else {
                            if (init.type === "UnaryExpression") {
                                value = this.sourceAtPos(init.start, init.end);
                                type = typeof init.argument?.value;
                            }
                            else if (init.type === "Identifier") {
                                // Handle non-literal defaults like variable references and global identifiers.
                                // Don't infer type, just preserve existing type annotation.
                                value = this.sourceAtPos(init.start, init.end);
                            }
                            else if (init.type === "MemberExpression") {
                                // Handle member expressions like Number.POSITIVE_INFINITY, Math.PI, etc.
                                value = this.sourceAtPos(init.start, init.end);
                                // Infer type as "number" for well-known numeric constants
                                if (this.isNumericConstant(init)) {
                                    type = "number";
                                }
                                // Otherwise, don't infer type, just preserve existing type annotation.
                            }
                            else {
                                value = init.raw;
                                type = init.value == null ? undefined : typeof init.value;
                            }
                        }
                    }
                    if (declaration_type === "FunctionDeclaration") {
                        // Don't store function body in value - it clutters documentation
                        // The type signature is what matters for API docs
                        value = undefined;
                        type = "() => any";
                        kind = "function";
                        isFunction = true;
                        isFunctionDeclaration = true;
                    }
                    let params;
                    let returnType;
                    if (node.leadingComments) {
                        const jsdoc_comment = ComponentParser.findJSDocComment(node.leadingComments);
                        if (jsdoc_comment) {
                            const comment = (0, comment_parser_1.parse)(ComponentParser.formatComment(jsdoc_comment.value), {
                                spacing: "preserve",
                            });
                            // Extract @type tag
                            const typeTag = comment[0]?.tags.find((t) => t.tag === "type");
                            if (typeTag)
                                type = this.aliasType(typeTag.type);
                            // Extract @param tags
                            const paramTags = comment[0]?.tags.filter((t) => t.tag === "param") ?? [];
                            if (paramTags.length > 0) {
                                params = paramTags
                                    .filter((tag) => !tag.name.includes(".")) // Exclude nested params like "options.expand"
                                    .map((tag) => ({
                                    name: tag.name,
                                    type: this.aliasType(tag.type),
                                    description: tag.description?.replace(PROPERTY_DESCRIPTION_REGEX, "").trim(),
                                    optional: tag.optional || false,
                                }));
                            }
                            // Extract @returns/@return tag
                            const returnsTag = comment[0]?.tags.find((t) => t.tag === "returns" || t.tag === "return");
                            if (returnsTag)
                                returnType = this.aliasType(returnsTag.type);
                            // Build description from comment description and non-param/non-type tags
                            const commentDescription = ComponentParser.assignValue(comment[0]?.description?.trim());
                            const additional_tags = comment[0]?.tags.filter((tag) => ![
                                "type",
                                "param",
                                "returns",
                                "return",
                                "extends",
                                "restProps",
                                "slot",
                                "event",
                                "typedef",
                            ].includes(tag.tag)) ?? [];
                            if (commentDescription || additional_tags.length > 0) {
                                description = commentDescription || "";
                                for (const tag of additional_tags) {
                                    description += `${description ? "\n" : ""}@${tag.tag}${tag.name ? ` ${tag.name}` : ""}${tag.description ? ` ${tag.description}` : ""}`;
                                }
                            }
                        }
                    }
                    if (!description && this.typedefs.has(type)) {
                        description = this.typedefs.get(type)?.description;
                    }
                    this.addProp(prop_name, {
                        name: prop_name,
                        kind,
                        description,
                        type,
                        value,
                        params,
                        returnType,
                        isFunction,
                        isFunctionDeclaration,
                        isRequired,
                        constant: kind === "const",
                        reactive: this.reactive_vars.has(prop_name),
                    });
                }
                if (node.type === "Comment") {
                    const data = node?.data?.trim() ?? "";
                    if (COMPONENT_COMMENT_REGEX.test(data)) {
                        this.componentComment = data.replace(COMPONENT_COMMENT_REGEX, "").replace(CARRIAGE_RETURN_REGEX, "");
                    }
                }
                if (node.type === "Slot") {
                    const slot_name = node.attributes.find((attr) => attr.name === "name")?.value[0].data;
                    const slot_props = node.attributes
                        .filter((attr) => attr.name !== "name")
                        .reduce((slot_props, { name, value }) => {
                        const slot_prop_value = {
                            value: undefined,
                            replace: false,
                        };
                        if (value === undefined)
                            return slot_props;
                        if (value[0]) {
                            const { type, expression, raw, start, end } = value[0];
                            if (type === "Text") {
                                slot_prop_value.value = raw;
                            }
                            else if (type === "AttributeShorthand") {
                                slot_prop_value.value = expression.name;
                                slot_prop_value.replace = true;
                            }
                            if (expression) {
                                if (expression.type === "Literal") {
                                    slot_prop_value.value = expression.value;
                                }
                                else if (expression.type !== "Identifier") {
                                    if (expression.type === "ObjectExpression" || expression.type === "TemplateLiteral") {
                                        slot_prop_value.value = this.sourceAtPos(start + 1, end - 1);
                                    }
                                    else {
                                        slot_prop_value.value = this.sourceAtPos(start, end);
                                    }
                                }
                            }
                        }
                        slot_props[name] = slot_prop_value;
                        return slot_props;
                    }, {});
                    const fallback = node.children
                        ?.map(({ start, end }) => this.sourceAtPos(start, end))
                        .join("")
                        .trim();
                    this.addSlot({
                        slot_name,
                        slot_props: JSON.stringify(slot_props, null, 2),
                        slot_fallback: fallback,
                    });
                }
                if (node.type === "EventHandler" && node.expression == null) {
                    if (parent !== undefined) {
                        // Track that this event is forwarded (we'll use this info later)
                        this.forwardedEvents.set(node.name, parent.name);
                        const existing_event = this.events.get(node.name);
                        // Check if this event has a JSDoc description
                        const description = this.eventDescriptions.get(node.name);
                        const event_description = description?.split("-").pop()?.trim();
                        if (!existing_event) {
                            // Add new forwarded event
                            this.events.set(node.name, {
                                type: "forwarded",
                                name: node.name,
                                element: parent.name,
                                description: event_description,
                            });
                        }
                        else if (existing_event.type === "forwarded" && event_description && !existing_event.description) {
                            // Event is already forwarded, just add the description
                            this.events.set(node.name, {
                                ...existing_event,
                                description: event_description,
                            });
                        }
                        // Note: if event is dispatched, we don't overwrite it here
                        // We'll handle @event JSDoc on forwarded events after the walk completes
                    }
                }
                if (parent?.type === "Element" && node.type === "Binding" && node.name === "this") {
                    const prop_name = node.expression.name;
                    const element_name = parent.name;
                    if (this.bindings.has(prop_name)) {
                        const existing_bindings = this.bindings.get(prop_name);
                        if (!existing_bindings.elements.includes(element_name)) {
                            this.bindings.set(prop_name, {
                                ...existing_bindings,
                                elements: [...existing_bindings.elements, element_name],
                            });
                        }
                    }
                    else {
                        this.bindings.set(prop_name, {
                            elements: [element_name],
                        });
                    }
                }
            },
        });
        if (dispatcher_name !== undefined) {
            for (const callee of callees) {
                if (callee.name === dispatcher_name) {
                    const event_name = callee.arguments[0]?.value;
                    const event_argument = callee.arguments[1];
                    const event_detail = event_argument?.value;
                    this.addDispatchedEvent({
                        name: event_name,
                        detail: event_detail,
                        has_argument: Boolean(event_argument),
                    });
                }
            }
        }
        // Post-process events: convert dispatched events from @event JSDoc to forwarded events
        // if they are actually forwarded and not dispatched via createEventDispatcher
        const actuallyDispatchedEvents = new Set();
        if (dispatcher_name !== undefined) {
            for (const callee of callees) {
                if (callee.name === dispatcher_name) {
                    actuallyDispatchedEvents.add(callee.arguments[0]?.value);
                }
            }
        }
        this.forwardedEvents.forEach((element, eventName) => {
            const event = this.events.get(eventName);
            // If event is marked as dispatched but is NOT actually dispatched, convert it to forwarded
            if (event && event.type === "dispatched" && !actuallyDispatchedEvents.has(eventName)) {
                const description = this.eventDescriptions.get(eventName);
                const event_description = description?.split("-").pop()?.trim();
                const forwardedEvent = {
                    type: "forwarded",
                    name: eventName,
                    element: element,
                    description: event_description,
                };
                // Preserve detail type if it was explicitly set in @event tag
                // Note: "null" is a valid explicit type (e.g., @event {null} eventname)
                // Only skip if detail is truly undefined or the string "undefined"
                if (event.detail !== undefined && event.detail !== "undefined") {
                    forwardedEvent.detail = event.detail;
                }
                this.events.set(eventName, forwardedEvent);
            }
        });
        return {
            props: ComponentParser.mapToArray(this.props).map((prop) => {
                if (this.bindings.has(prop.name)) {
                    return {
                        ...prop,
                        type: "null | " +
                            this.bindings
                                .get(prop.name)
                                ?.elements.sort()
                                .map((element) => (0, element_tag_map_1.getElementByTag)(element))
                                .join(" | "),
                    };
                }
                return prop;
            }),
            moduleExports: ComponentParser.mapToArray(this.moduleExports),
            slots: ComponentParser.mapToArray(this.slots)
                .map((slot) => {
                try {
                    const slot_props = JSON.parse(slot.slot_props);
                    const new_props = [];
                    for (const key of Object.keys(slot_props)) {
                        if (slot_props[key].replace && slot_props[key].value !== undefined) {
                            slot_props[key].value = this.props.get(slot_props[key].value)?.type;
                        }
                        if (slot_props[key].value === undefined)
                            slot_props[key].value = "any";
                        new_props.push(`${key}: ${slot_props[key].value}`);
                    }
                    const formatted_slot_props = new_props.length === 0 ? "Record<string, never>" : `{ ${new_props.join(", ")} }`;
                    return { ...slot, slot_props: formatted_slot_props };
                }
                catch (_e) {
                    return slot;
                }
            })
                .sort((a, b) => {
                if (a.name < b.name)
                    return -1;
                if (a.name > b.name)
                    return 1;
                return 0;
            }),
            events: ComponentParser.mapToArray(this.events),
            typedefs: ComponentParser.mapToArray(this.typedefs),
            generics: this.generics,
            rest_props: this.rest_props,
            extends: this.extends,
            componentComment: this.componentComment,
            contexts: ComponentParser.mapToArray(this.contexts),
        };
    }
}
exports.default = ComponentParser;
