// sveld/lib/resolve-alias.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearConfigCache = clearConfigCache;
exports.resolvePathAliasAbsolute = resolvePathAliasAbsolute;
exports.resolvePathAlias = resolvePathAlias;
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const path_1 = require("./path");
const configCache = new Map();
const COMMENT_PATTERN = /\/\*[\s\S]*?\*\/|\/\/.*/g;
const REGEX_SPECIAL_CHARS = /[.+?^${}()|[\]\\]/g;
function clearConfigCache() {
    configCache.clear();
}
// Find nearest tsconfig.json/jsconfig.json starting from a directory
function findConfig(startDir) {
    let dir = startDir;
    const root = (0, node_path_1.resolve)(dir, "/");
    while (dir !== root) {
        for (const configName of ["tsconfig.json", "jsconfig.json"]) {
            const configPath = (0, node_path_1.join)(dir, configName);
            if ((0, node_fs_1.existsSync)(configPath)) {
                return configPath;
            }
        }
        dir = (0, node_path_1.dirname)(dir);
    }
    return null;
}
function parseConfig(configPath) {
    if (configCache.has(configPath)) {
        return configCache.get(configPath) ?? null;
    }
    try {
        const content = (0, node_fs_1.readFileSync)(configPath, "utf-8");
        const jsonContent = content.replace(COMMENT_PATTERN, "");
        const config = JSON.parse(jsonContent);
        if (config.extends) {
            const baseConfigPath = (0, node_path_1.isAbsolute)(config.extends) ? config.extends : (0, node_path_1.resolve)((0, node_path_1.dirname)(configPath), config.extends);
            const fullBaseConfigPath = baseConfigPath.endsWith(".json") ? baseConfigPath : `${baseConfigPath}.json`;
            if ((0, node_fs_1.existsSync)(fullBaseConfigPath)) {
                const baseConfig = parseConfig(fullBaseConfigPath);
                if (baseConfig) {
                    config.compilerOptions = {
                        ...baseConfig.compilerOptions,
                        ...config.compilerOptions,
                        paths: {
                            ...baseConfig.compilerOptions?.paths,
                            ...config.compilerOptions?.paths,
                        },
                    };
                }
            }
        }
        configCache.set(configPath, config);
        return config;
    }
    catch {
        configCache.set(configPath, null);
        return null;
    }
}
// Resolve a path alias to an absolute file system path for reading files.
// Returns the original path if it cannot be resolved.
function resolvePathAliasAbsolute(importPath, fromDir) {
    if (importPath.startsWith(".") || importPath.startsWith("/")) {
        return importPath;
    }
    const configPath = findConfig(fromDir);
    if (!configPath) {
        return importPath;
    }
    const config = parseConfig(configPath);
    if (!config?.compilerOptions?.paths) {
        return importPath;
    }
    const { baseUrl = ".", paths } = config.compilerOptions;
    const configDir = (0, node_path_1.dirname)(configPath);
    const resolvedBaseUrl = (0, node_path_1.resolve)(configDir, baseUrl);
    // Try to match against path patterns
    for (const [pattern, mappings] of Object.entries(paths)) {
        // Convert TS path pattern to regex
        // e.g., "$lib/*" -> /^\$lib\/(.*)$/
        // e.g., "$lib" -> /^\$lib$/
        // e.g., "@components/*" -> /^@components\/(.*)$/
        // Escape special regex chars but keep * for replacement
        const escapedPattern = pattern
            .split("*")
            .map((part) => part.replace(REGEX_SPECIAL_CHARS, "\\$&"))
            .join("(.*)");
        const regex = new RegExp(`^${escapedPattern}$`);
        const match = importPath.match(regex);
        if (match) {
            // Use the first mapping (TypeScript uses the first match)
            const mapping = mappings[0];
            if (!mapping)
                continue;
            // Replace * in mapping with captured groups
            let resolvedPath = mapping;
            for (let i = 1; i < match.length; i++) {
                resolvedPath = resolvedPath.replace("*", match[i]);
            }
            // Resolve relative to baseUrl
            const fullPath = (0, node_path_1.resolve)(resolvedBaseUrl, resolvedPath);
            return fullPath;
        }
    }
    return importPath;
}
// Resolve a path alias and convert to a relative path from fromDir.
// This is used for storing paths in the exports object for output generation.
// Returns the original path if it cannot be resolved.
function resolvePathAlias(importPath, fromDir) {
    if (importPath.startsWith(".") || importPath.startsWith("/")) {
        return importPath;
    }
    const absolutePath = resolvePathAliasAbsolute(importPath, fromDir);
    if (absolutePath === importPath) {
        return importPath;
    }
    let relativePath = (0, node_path_1.relative)(fromDir, absolutePath);
    // Normalize path separators to forward slashes for consistency across platforms
    relativePath = (0, path_1.normalizeSeparators)(relativePath);
    return relativePath.startsWith(".") ? relativePath : `./${relativePath}`;
}
