// sveld/lib/rollup-plugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = pluginSveld;
exports.generateBundle = generateBundle;
exports.writeOutput = writeOutput;
const node_fs_1 = require("node:fs");
const promises_1 = require("node:fs/promises");
const node_path_1 = require("node:path");
const compiler_1 = require("svelte/compiler");
const svelte_preprocess_1 = require("svelte-preprocess");
const tinyglobby_1 = require("tinyglobby");
const ComponentParser_1 = __importDefault(require("./ComponentParser"));
const get_svelte_entry_1 = require("./get-svelte-entry");
const parse_exports_1 = require("./parse-exports");
const path_1 = require("./path");
const writer_json_1 = __importDefault(require("./writer/writer-json"));
const writer_markdown_1 = __importDefault(require("./writer/writer-markdown"));
const writer_ts_definitions_1 = __importDefault(require("./writer/writer-ts-definitions"));
const STYLE_TAG_REGEX = /<style.+?<\/style>/gims;
const HYPHEN_REGEX = /-/g;
function pluginSveld(opts) {
    let result;
    let input;
    return {
        name: "plugin-sveld",
        buildStart() {
            input = (0, get_svelte_entry_1.getSvelteEntry)(opts?.entry);
        },
        async generateBundle() {
            if (input != null) {
                result = await generateBundle(input, opts?.glob === true);
            }
        },
        writeBundle() {
            if (input != null)
                writeOutput(result, opts || {}, input);
        },
    };
}
async function generateBundle(input, glob) {
    const isFile = (0, node_fs_1.lstatSync)(input).isFile();
    const dir = isFile ? (0, node_path_1.dirname)(input) : input;
    // Only parse exports if input is a file
    let exports = {};
    if (isFile) {
        const entry = (0, node_fs_1.readFileSync)(input, "utf-8");
        exports = (0, parse_exports_1.parseExports)(entry, dir);
    }
    const allComponents = { ...exports };
    if (glob) {
        for (const file of (0, tinyglobby_1.globSync)([`${dir}/**/*.svelte`])) {
            const moduleName = (0, node_path_1.parse)(file).name.replace(HYPHEN_REGEX, "");
            const source = (0, path_1.normalizeSeparators)(`./${(0, node_path_1.relative)(dir, file)}`);
            if (exports[moduleName]) {
                exports[moduleName].source = source;
            }
            if (allComponents[moduleName]) {
                allComponents[moduleName].source = source;
            }
            else {
                allComponents[moduleName] = { source, default: false };
            }
        }
    }
    const components = new Map();
    const allComponentsForTypes = new Map();
    const exportEntries = Object.entries(exports);
    const allComponentEntries = Object.entries(allComponents);
    // Process exported components (for metadata/JSON/Markdown)
    const componentPromises = exportEntries.map(async ([exportName, entry]) => {
        const filePath = entry.source;
        const { ext, name } = (0, node_path_1.parse)(filePath);
        let moduleName = exportName;
        if (exportEntries.length === 1 && exportName === "default") {
            moduleName = name;
        }
        if (ext === ".svelte") {
            const source = await (0, promises_1.readFile)((0, node_path_1.resolve)(dir, filePath), "utf-8");
            const { code: processed } = await (0, compiler_1.preprocess)(source, [(0, svelte_preprocess_1.typescript)(), (0, svelte_preprocess_1.replace)([[STYLE_TAG_REGEX, ""]])], {
                filename: (0, node_path_1.basename)(filePath),
            });
            const parser = new ComponentParser_1.default();
            const parsed = parser.parseSvelteComponent(processed, {
                moduleName,
                filePath,
            });
            return {
                moduleName,
                filePath,
                ...parsed,
            };
        }
        return null;
    });
    // Process all components (for .d.ts generation)
    const allComponentPromises = allComponentEntries.map(async ([exportName, entry]) => {
        const filePath = entry.source;
        const { ext, name } = (0, node_path_1.parse)(filePath);
        let moduleName = exportName;
        if (allComponentEntries.length === 1 && exportName === "default") {
            moduleName = name;
        }
        if (ext === ".svelte") {
            const source = await (0, promises_1.readFile)((0, node_path_1.resolve)(dir, filePath), "utf-8");
            const { code: processed } = await (0, compiler_1.preprocess)(source, [(0, svelte_preprocess_1.typescript)(), (0, svelte_preprocess_1.replace)([[STYLE_TAG_REGEX, ""]])], {
                filename: (0, node_path_1.basename)(filePath),
            });
            const parser = new ComponentParser_1.default();
            const parsed = parser.parseSvelteComponent(processed, {
                moduleName,
                filePath,
            });
            return {
                moduleName,
                filePath,
                ...parsed,
            };
        }
        return null;
    });
    const [results, allResults] = await Promise.all([Promise.all(componentPromises), Promise.all(allComponentPromises)]);
    for (const result of results) {
        if (result) {
            components.set(result.moduleName, result);
        }
    }
    for (const result of allResults) {
        if (result) {
            allComponentsForTypes.set(result.moduleName, result);
        }
    }
    return {
        exports,
        components,
        allComponentsForTypes,
    };
}
function writeOutput(result, opts, input) {
    const inputDir = (0, node_path_1.dirname)(input);
    if (opts?.types !== false) {
        // Use allComponentsForTypes to generate .d.ts for all discovered components
        (0, writer_ts_definitions_1.default)(result.allComponentsForTypes, {
            outDir: "types",
            preamble: "",
            ...opts?.typesOptions,
            exports: result.exports,
            inputDir,
        });
    }
    if (opts?.json) {
        // Use components (exported only) for JSON metadata
        (0, writer_json_1.default)(result.components, {
            outFile: "COMPONENT_API.json",
            ...opts?.jsonOptions,
            input,
            inputDir,
        });
    }
    if (opts?.markdown) {
        // Use components (exported only) for Markdown documentation
        (0, writer_markdown_1.default)(result.components, {
            outFile: "COMPONENT_INDEX.md",
            ...opts?.markdownOptions,
        });
    }
}
